pkg/ingester/limiting.go

```diff
@@ -0,0 +1,239 @@
+		var name string
+		var val uint64
+		if _, err := fmt.Fscanf(memR, "%s %d\n", &name, &val); err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+
+			return 0, 0, errors.Wrapf(err, "failed scanning %s", memPath)
+		}
+
+		if name == "rss" || name == "cache" {
+			memUtil += val
+			numMemStats++
+			if numMemStats == 2 {
+				break
+			}
+		}
+	}
+	if numMemStats != 2 {
+		return 0, 0, fmt.Errorf("failed scanning %s", memPath)
+	}
+
+	// For reference, see https://www.kernel.org/doc/Documentation/cgroup-v1/cpuacct.txt
+	const cpuPath = "/sys/fs/cgroup/cpu/cpuacct.usage"
+	cpuR, err := readFileNoStat(cpuPath)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer cpuR.Close()
+	// CPU time in nanoseconds
+	var cpuTime uint64
+	if _, err := fmt.Fscanf(cpuR, "%d", &cpuTime); err != nil {
+		return 0, 0, errors.Wrapf(err, "failed scanning %s", cpuPath)
+	}
+
+	return float64(cpuTime) / 1e09, memUtil, nil
+}
+
+func scanProcFS() (float64, uint64, error) {
+	p, err := procfs.Self()
+	if err != nil {
+		return 0, 0, errors.Wrap(err, "failed to get process info")
+	}
+	ps, err := p.Stat()
+	if err != nil {
+		return 0, 0, errors.Wrap(err, "failed to get process stats")
+	}
+
+	return ps.CPUTime(), uint64(ps.ResidentMemory()), nil
+}
+
+func (i *Ingester) updateResourceUtilization() {
+	if !i.cfg.UtilizationBasedLimitingEnabled {
+		return
+	}
+
+	lastUpdate := i.lastResourceUtilizationUpdate.Load()
+
+	var method string
+	cpuTime, memUtil, err := scanCgroupV2()
+	if err != nil && !os.IsNotExist(err) {
+		level.Warn(i.logger).Log("msg", "failed to get CPU and memory stats from cgroup v2", "err", err.Error())
+		return
+	}
+	if err == nil {
+		method = "cgroup v2"
+	} else {
+		// cgroup v2 not detected, try v1
+		cpuTime, memUtil, err = scanCgroupV1()
+		if err != nil && !os.IsNotExist(err) {
+			level.Warn(i.logger).Log("msg", "failed to get CPU and memory stats from cgroup v1", "err", err.Error())
+			return
+		}
+		if err == nil {
+			method = "cgroup v1"
+		} else {
+			// cgroup not detected, fall back to /proc
+			cpuTime, memUtil, err = scanProcFS()
+			if err != nil {
+				level.Warn(i.logger).Log("msg", "failed to get CPU and memory stats from /proc", "err", err.Error())
+				return
+			}
+			method = "/proc"
+		}
+	}
+
+	i.memoryUtilization.Store(memUtil)
+
+	now := time.Now().UTC()
+
+	lastCPUTime := i.lastCPUTime.Load()
+	i.lastCPUTime.Store(cpuTime)
+
+	i.lastResourceUtilizationUpdate.Store(now)
+
+	if lastUpdate.IsZero() {
+		return
+	}
+
+	cpuUtil := (cpuTime - lastCPUTime) / now.Sub(lastUpdate).Seconds()
+	i.movingAvg.Add(cpuUtil)
+	cpuA := i.movingAvg.Value()
+	i.cpuUtilization.Store(cpuA)
+
+	level.Debug(i.logger).Log("msg", "process resource utilization", "method", method, "memory_utilization", memUtil,
+		"smoothed_cpu_utilization", cpuA, "raw_cpu_utilization", cpuUtil)
+}
+
+// checkReadOverloaded checks whether the ingester read path is overloaded wrt. CPU and/or memory.
+func (i *Ingester) checkReadOverloaded() error {
```

Comment:
Why do we have to do here all the math to figure out if the read-path is overloaded? Can't be pre-computed by `updateResourceUtilization()`? Basically `updateResourceUtilization()` could just compute if the read path is overloaded, rather than the actual utilization. This way the cost to check if the read-path is overloaded is going to be a function of the frequency at which you run `updateResourceUtilization()` (predictable) rather than the query load (unpredicatable).

Reply:
I've pushed a revision where I pre-compute during the periodic update. PTAL.